src/pcp.rs

```diff
@@ -0,0 +1,596 @@
+use crate::field::FieldElement;
+use crate::fp::log2;
+use crate::polynomial::{poly_deg, poly_eval, poly_mul};
+
+pub mod gadgets;
+pub mod types;
+
+/// Errors propagagted by methods in this module.
+#[derive(Debug, PartialEq, thiserror::Error)]
+pub enum PcpError {
+    /// The caller of an arithmetic circuit provided the wrong number of inputs. This error may
+    /// occur when evaluating a validity circuit or gadget.
+    #[error("wrong number of inputs to arithmetic circuit")]
+    CircuitInLen,
+
+    /// This error is returned by `collect` if the input slice is empty.
+    #[error("collect requires at least one input")]
+    CollectInLen,
+
+    /// This error is returned by `collect` if the two or more verifier shares have different
+    /// gadget arities.
+    #[error("collect inputs have mismatched gadget arity")]
+    CollectGadgetInLenMismatch,
+
+    /// Returned if an FFT operation propagates an error.
+    #[error("FFT error")]
+    Fft(#[from] FftError),
+
+    /// When evaluating a gadget on polynomials, this error is returned if the input polynomials
+    /// don't all have the same length.
+    #[error("gadget called on polynomials with different lengths")]
+    GadgetPolyInLen,
+
+    /// When evaluating a gadget on polynomials, this error is returned if the slice allocated for
+    /// the output polynomial is too small.
+    #[error("sliice allocated for gadget output is too small")]
+    GadgetPolyOutLen,
+
+    /// The proof string is either too large or too short for the given type.
+    #[error("attempted query on proof with invalid length")]
+    QueryProofLen,
+}
+
+/// A data type. Implementations of this trait specify an arithmetic circuit that determines
+/// whether a given value is valid.
+pub trait Type<F, G>: Sized
+where
+    F: FieldElement,
+    G: Gadget<F>,
+{
+    /// Evalauates the validity circuit on the given input (i.e., `self`) and returns the output.
+    /// Slice `rand` is the random input consumed by the validity circuit.
+    ///
+    /// ```
+    /// use prio::pcp::types::Boolean;
+    /// use prio::pcp::Type;
+    /// use prio::field::{FieldElement, Field64};
+    ///
+    /// type F = Field64;
+    /// type T = Boolean<F>;
+    ///
+    /// let x = T::new(false);
+    ///
+    /// let mut rand: Vec<F> = Vec::with_capacity(x.valid_rand_len());
+    /// for _ in 0..rand.len() {
+    ///     rand.push(F::rand());
+    /// }
+    ///
+    /// let v = x.valid(&mut x.gadget(0), &rand).unwrap();
+    /// assert_eq!(v, F::zero());
+    /// ```
+    fn valid(&self, g: &mut dyn GadgetCallOnly<F>, rand: &[F]) -> Result<F, PcpError>;
+
+    /// Returns a reference to the underlying data.
+    fn as_slice(&self) -> &[F];
+
+    /// Constructs an instance of this type from the given data without checking whether the data
+    /// is valid. This method takes ownership of `data`.
+    fn clone_with(&self, data: Vec<F>) -> Self;
```

Comment:
FWIW, this function is useful for testing, but not really very useful in the public API. I think we should tag this with `#[cfg(test)]`.

Reply:
Per our discussion in Slack last week, I think this should become `fn new_with(data: Vec<F>) -> Self;`. We agreed that if we ever do encounter a `Value` impl that can't be instantiated without an existing value, we can break this semi-private interface. Further I think we should avoid using "clone" in the name of this function because I think that implies `std::clone::Clone` semantics to the reader.

Reply:
Changed to `new_with`.